src/charm.py

```diff
@@ -1,35 +1,180 @@
 #!/usr/bin/env python3
-# Copyright 2025 Iurii Kondrakov
+# Copyright 2025 Canonical Ltd.
 # See LICENSE file for licensing details.
-#
-# Learn more at: https://juju.is/docs/sdk
 
-"""Charm the service.
-
-Refer to the following tutorial that will help you
-develop a new k8s charm using the Operator Framework:
-
-https://juju.is/docs/sdk/create-a-minimal-kubernetes-charm
-"""
+"""The Velero Charm."""
 
 import logging
+from typing import Type, Union
 
 import ops
+from lightkube import ApiError, Client
+from lightkube.resources.rbac_authorization_v1 import ClusterRole
+
+from config import (
+    USE_NODE_AGENT_CONFIG_KEY,
+    VELERO_AWS_PLUGIN_CONFIG_KEY,
+    VELERO_AZURE_PLUGIN_CONFIG_KEY,
+    VELERO_IMAGE_CONFIG_KEY,
+    VELERO_PATH,
+)
+from velero import Velero, VeleroError
 
-# Log messages can be retrieved using juju debug-log
 logger = logging.getLogger(__name__)
 
 
+class WithStatusError(Exception):
+    """Base class of exceptions for when the raiser has an opinion on the charm status."""
+
+    def __init__(
+        self,
+        msg: str,
+        status_type: Type[
+            Union[ops.ActiveStatus, ops.WaitingStatus, ops.BlockedStatus, ops.MaintenanceStatus]
+        ],
+    ):
+        super().__init__(str(msg))
+        self.msg = str(msg)
+        self.status_type = status_type
+
+    @property
+    def status(self):
+        """Return an instance of self.status_type, instantiated with this exception's message."""
+        return self.status_type(self.msg)
+
+
 class VeleroOperatorCharm(ops.CharmBase):
     """Charm the service."""
 
+    _stored = ops.StoredState()
+
     def __init__(self, framework: ops.Framework):
         super().__init__(framework)
+        self._stored.set_default(
+            storage_provider_attached=None,
+        )
+
+        # Lightkube client needed for interacting with the Kubernetes cluster
+        self.lightkube_client = None
+
+        try:
+            self._validate_config()
```

Comment:
validating config on each hook calls seem not that efficient

Reply:
In the end we went with using the Data Platforms pydantic logic for handling the config, and having a python object with properties for the configs (with correct types as well).

This is a pattern we do with other charms as well, so I suggest that we move with this one for now.